src/librbd/internal.cc

```diff
@@ -1566,6 +1580,11 @@ bool compare_by_name(const child_info_t& c1, const child_info_t& c2)
       return r;
     }
 
+    if (trash_spec.state == cls::rbd::TRASH_IMAGE_STATE_PENDING) {
```

Comment:
Note that there is still a (smaller) race condition where you can start to restore and then delete the image from the trash. I guess to fully solve it would require multiple states (moving, removing, restoring) and the ```trash_state_update``` method should take an expected-state argument (or just require the state to be normal or match).

Reply:
@Songweibin I think this last point is still valid in that there is a small race condition where two processes can start performing actions and race.